mlos_bench/mlos_bench/storage/sql_experiment.py

```diff
@@ -0,0 +1,209 @@
+                ).order_by(
+                    self._schema.trial.c.trial_id.asc(),
+                )
+            )
+            for trial in cur_trials.fetchall():
+                tunables = self._get_params(
+                    conn, self._schema.config_param, config_id=trial.config_id)
+                configs.append(tunables)
+                scores.append(float(trial.metric_value))
+            return (configs, scores)
+
+    @staticmethod
+    def _get_params(conn: Connection, table: Table, **kwargs: Any) -> Dict[str, Any]:
+        cur_params = conn.execute(table.select().where(*[
+            column(key) == val for (key, val) in kwargs.items()]))
+        return {row.param_id: row.param_value for row in cur_params.fetchall()}
+
+    @staticmethod
+    def _save_params(conn: Connection, table: Table,
+                     params: Dict[str, Any], **kwargs: Any) -> None:
+        conn.execute(table.insert(), [
+            {
+                **kwargs,
+                "param_id": key,
+                "param_value": None if val is None else str(val)
+            }
+            for (key, val) in params.items()
+        ])
+
+    def pending_trials(self) -> Iterator[Storage.Trial]:
+        _LOG.info("Retrieve pending trials for: %s", self._experiment_id)
+        with self._engine.connect() as conn:
+            cur_trials = conn.execute(self._schema.trial.select().where(
+                self._schema.trial.c.exp_id == self._experiment_id,
+                self._schema.trial.c.ts_end.is_(None)
+            ))
+            for trial in cur_trials.fetchall():
+                tunables = self._get_params(
+                    conn, self._schema.config_param,
+                    config_id=trial.config_id)
+                config = self._get_params(
+                    conn, self._schema.trial_param,
+                    exp_id=self._experiment_id, trial_id=trial.trial_id)
+                yield Trial(
+                    engine=self._engine,
+                    schema=self._schema,
+                    # Reset .is_updated flag after the assignment:
+                    tunables=self._tunables.copy().assign(tunables).reset(),
+                    experiment_id=self._experiment_id,
+                    trial_id=trial.trial_id,
+                    config_id=trial.config_id,
+                    opt_target=self._opt_target,
+                    config=config,
+                )
+
+    def _get_config_id(self, conn: Connection, tunables: TunableGroups) -> int:
+        """
+        Get the config ID for the given tunables. If the config does not exist,
+        create a new record for it.
+        """
+        config_hash = hashlib.sha256(str(tunables).encode('utf-8')).hexdigest()
+        cur_config = conn.execute(self._schema.config.select().where(
+            column("config_hash") == config_hash
+        )).fetchone()
+        if cur_config is not None:
+            return int(cur_config.config_id)  # mypy doesn't know it's always int
```

Comment:
Hmm, that's too bad.  I would have hoped that that came from the schema details that we setup in sqlalchemy.

Reply:
yeah, I believe that SQLAlchemy just cannot guarantee that it can return the right type for any backend it supports. Anyway, it should be a safe cast here